File: 221 Maximal Square.cpp
#include <vector>
using namespace std;
/*
Maximal Square
https://leetcode.com/problems/maximal-square/
*/
int min3(int a,int b,int c){
    int m = a < b ? a : b;
    return m < c ? m : c;
}

int maximalSquare(vector<vector<char> >& mat){
    size_t row,col,i,j;
    int maxv,v;
    
    row = mat.size();
    if(row == 0) return 0;
    
    col = mat[0].size();
    
    maxv = 0;
    for(i=0;i<row;i++){
        if(mat[i][0] == '1'){
            maxv = 1;
            break;
        }
    }
    if(maxv != 1){
        for(i=1;i<col;i++){
            if(mat[0][i] == '1'){
                maxv = 1;
                break;
            }
        }
    }
    /*
    Use the input as buffer
    */
    for(i=1;i<row;i++){
        for(j=1;j<col;j++){
            if(mat[i][j] != '0'){
                v = min3(mat[i-1][j]-'0',mat[i-1][j-1]-'0',mat[i][j-1]-'0')+1;
                mat[i][j] = v + '0';
                maxv = maxv >= v ? maxv : v ;
            }
        }
    }
    return maxv * maxv;
    
int main(){
    return 0;
}

File: 221 Maximal Square.py
'''
221. Maximal Square
https://leetcode.com/problems/maximal-square

Given a 2D binary matrix filled with 0's and 1's, find the largest square containing only 1's and return its area.

Example:

Input: 

1 0 1 0 0
1 0 1 1 1
1 1 1 1 1
1 0 0 1 0

Output: 4
'''
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        if not matrix or not matrix[0]:
            return 0
        
        rowCount = len(matrix)
        colCount = len(matrix[0])
        
        # dp[i][j] is size of maximal square whose right-down point is (i,j)
        dp = [[0] * colCount for _ in range(rowCount)]
        maxSize = 0
        
        for i in range(rowCount):
            for j in range(colCount):
                if matrix[i][j] == '1':
                    if i > 0 and j > 0:
                        dp[i][j] = min(dp[i-1][j], dp[i][j-1], dp[i-1][j-1]) + 1
                    else:
                        dp[i][j] = 1
                    maxSize = max(maxSize, dp[i][j])
        
        return maxSize * maxSize
